File: csharp-dotnet/src/Architecture.Core/Domain/Entities/IEntity.cs
namespace Architecture.Core.Domain.Entities;

/// <summary>
/// Defines the contract for domain entities with identity-based equality.
/// Entities are objects that have a distinct identity that runs through time and different representations.
/// </summary>
/// <typeparam name="TId">The type of the entity's identifier, constrained to reference types.</typeparam>
public interface IEntity<TId>
    where TId : class
{
    /// <summary>
    /// Gets the unique identifier for this entity.
    /// The identity distinguishes this entity from all others and remains constant throughout the entity's lifetime.
    /// </summary>
    TId Id { get; }
}

File: typescript-nodejs/src/domain/interfaces/i-entity.ts
/**
 * Interface for domain entities.
 * Entities are objects that have a distinct identity that runs through time and different representations.
 * They are defined primarily by their identity, not their attributes.
 *
 * @template TId The type of the entity's identifier
 */
export interface IEntity<TId extends object> {
  /** The unique identifier of the entity */
  readonly id: TId;

  /**
   * Determines equality based on identity rather than attribute values.
   * @param other The other entity to compare with
   * @returns True if entities have the same identity
   */
  equals(other: unknown): boolean;

  /**
   * Gets the hash code for the entity based on its identity.
   * @returns Hash code for the entity
   */
  getHashCode(): number;
}